### PR TITLE
function to jsonize prophy structs

### PR DESCRIPTION
**Pull request by mareklabonarski**

### CODE REVIEW

File: prophy/tests/test_utils.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    s.fixed_array[1].im = 22
+    s.fixed_array[2].re = 3
+    s.fixed_array[2].im = -33
+    s.dynamic.add()
+    s.dynamic[0].re = 45
+    s.dynamic[0].im = 55
+
+    result = jsonize(s, False)
+    from pprint import pprint
+    pprint(result)
```
Comment (kamichal) [inline prophy/tests/test_utils.py]: no need to `pprint`

File: prophy/tests/test_utils.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            {'im': -33, 're': 3}
+        ],
+        'optional_element': 666,
+        'that_enum': 2,
+        'that_union': 12345
+    }
+
+    ordered_result = jsonize(s, True)
+
+    pprint(ordered_result)
```
Comment (kamichal) [inline prophy/tests/test_utils.py]: saa

File: prophy/tests/test_utils.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        ('that_union', 12345),
+        ('optional_element', 666),
+        ('fixed_array', [
+            [('re', 0), ('im', 0)], [('re', 2), ('im', 22)], [('re', 3), ('im', -33)],
+        ]),
+        ('ext_size', None),
+        ('dynamic', [
+            [('re', 45), ('im', 55)]
+        ]),
+    ]
```
Comment (kamichal) [inline prophy/tests/test_utils.py]: Please add tests for serializing pure python's `list`, `dict` and `tuple` also test empty arguments for each case. Check strings with `unicode` characters and `bytes` strings with non ASCII characters (e.g. with each of '\x00\x01\x0a\xaf\xff').

File: prophy/tests/test_utils.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    s.that_union.other = 12345
+    s.optional_element = True
+    s.optional_element = 666
+    s.fixed_array[1].re = 2
+    s.fixed_array[1].im = 22
+    s.fixed_array[2].re = 3
+    s.fixed_array[2].im = -33
+    s.dynamic.add()
+    s.dynamic[0].re = 45
+    s.dynamic[0].im = 55
```
Comment (kamichal) [inline prophy/tests/test_utils.py:100]: Please move creation of this `s` object to a fixture and split this test into two: `ordered = [True, False]`

File: prophy/utils.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        elif getattr(struct_, '__getitem__', None):
+            return [jsonize(item, ordered) for item in struct_]
+
+        else:
+            return struct_
+
+    except Exception as e:
+        error_msg = 'Failed to dictionize {}: {}: {}'.format(type(struct_).__name__, type(e).__name__, e)
+        print(error_msg)
+        return error_msg
```
Comment (kamichal) [inline prophy/utils.py]: Please make sure that `coverage` does not decrease after adding this code.

File: prophy/utils.py [hunk 6]
```diff
@@ -24,19 +23,18 @@ def jsonize(struct_, ordered=True):
...
         elif isinstance(struct_, prophy.composite.union):
-            disctiminated = next((dsc[0] for dsc in struct_._descriptor if dsc[2] == struct_.discriminator), None)
-            assert disctiminated is not None, "Failed to get currently discriminated union field"
-            return jsonize(getattr(struct_, disctiminated, None), ordered)
+            discriminated = next((dsc.name for dsc in struct_._descriptor if dsc.discriminator == struct_.discriminator), None)
+            assert discriminated is not None, "Failed to get currently discriminated union field"
+            return jsonize(getattr(struct_, discriminated, None), ordered)
 
-        elif isinstance(struct_, six.string_types):
+        elif isinstance(struct_, (basestring, str)):
```
Comment (kamichal) [inline prophy/utils.py]: What's wrong with `six.string_types`?

File: prophy/tests/test_utils.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+    class cint16_t(prophy.with_metaclass(prophy.struct_generator, prophy.struct)):
+        _descriptor = [
+            ('re', prophy.i16),
+            ('im', prophy.i16),
+        ]
+
+    class TheUnion(prophy.with_metaclass(prophy.union_generator, prophy.union)):
+        _descriptor = [
+            ('field_with_a_long_name', cint16_t, (1, 1)),
```
Comment (kamichal) [inline prophy/tests/test_utils.py]: `(1, 1)`? And it works? Seems we need to tighten argument's validation. I've never seen seen it before.

File: prophy/six.py [hunk 8]
```diff
@@ -6,6 +6,7 @@
     long = long
     from itertools import ifilter
     xrange = xrange
+    string_types = (basestring, )
```
Comment (kamichal) [inline prophy/six.py]: string types in meaning of text in python2 are `(str, unicode)`.
`basestring` is way too broad, raw `bytes` are also its sublasses:
```
>>> isinstance(b"\x02\x12", basestring)
True
```

Comment (kamichal) [reply prophy/six.py]: Oh, I'm sorry, It's just my misunderstanding:
https://docs.python.org/2/library/functions.html#basestring

File: prophy/utils.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+
+def _try_to_get_attr(obj, name):
+    try:
+        return getattr(obj, name)
+    except AttributeError as e:
+        return str(e)
+
+
+def jsonize(struct_, ordered=True):
```
Comment (aurzenligl) [inline prophy/utils.py:14]: Did you give yaml a thought? It seems to be "best suited for configuration" [1], and this is rather configuration (schema/typesystem of messages) than data (actual message objects with values). It's a little lighter than json for human eyes, allows comments, but might be less popular in terms of available parsers, etc.
```
>>> print(json.dumps(data, indent=True))
{
 "foo": {
  "x": [
   1,
   2,
   3
  ],
  "y": "uhishfusd",
  "z": 1203
 }
}
>>> print(yaml.dump(data))
foo:
  x: [1, 2, 3]
  y: uhishfusd
  z: 1203
```
Can you point out disadvantages of yaml that I fail to see while writing this? :)

[1] https://www.json2yaml.com/yaml-vs-json

File: prophy/utils.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    """
+        It returns a python dict/list but it supposed to be fully json-serializable.
+        If ordered it reflects fields order, i.e. gives a list of name-value pairs and a dict otherwise.
+    """
+    try:
+        if isinstance(struct_, prophy.composite.struct):
+            if ordered:
+                return [(dsc.name, jsonize(getattr(struct_, dsc.name, None), ordered)) for dsc in struct_._descriptor]
```
Comment (aurzenligl) [inline prophy/utils.py]: You can generalize lines which generate dictionaries and lists of 2-tuples this way (:
```
>>> container = list
>>> container(((x, x) for x in [1,2,3]))
[(1, 1), (2, 2), (3, 3)]
>>> container = dict
>>> container(((x, x) for x in [1,2,3]))
{1: 1, 2: 2, 3: 3}
```

File: prophy/utils.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+        elif getattr(struct_, '__getitem__', None):
+            return [jsonize(item, ordered) for item in struct_]
+
+        else:
+            return struct_
+
+    except Exception as e:
+        error_msg = 'Failed to dictionize {}: {}: {}'.format(type(struct_).__name__, type(e).__name__, e)
+        print(error_msg)
```
Comment (aurzenligl) [inline prophy/utils.py]: Printing directly to stdout doesn't seem like a good idea, especially in a library. It can be an exception rethrow (see how similar error reporting cases are handled now).

File: prophy/utils.py [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+            if ordered:
+                return [(dsc.name, jsonize(getattr(struct_, dsc.name, None), ordered)) for dsc in struct_._descriptor]
+            else:
+                return {dsc.name: jsonize(getattr(struct_, dsc.name, None), ordered) for dsc in struct_._descriptor}
+
+        elif isinstance(struct_, prophy.composite.union):
+            discriminated = next(
+                (dsc.name for dsc in struct_._descriptor if dsc.discriminator == struct_.discriminator), None
+            )
+            assert discriminated is not None, "Failed to get currently discriminated union field"
```
Comment (aurzenligl) [inline prophy/utils.py]: If this is production code (which is unit tested) - consider raise. `assert` is supposed to catch unexpected mistakes developer, not user, made. Anyway, why not just `assert discriminated, "Fail(...)`?

**Review by aurzenligl — COMMENTED** [review]
Just a bunch of comments; it would be nice if purpose of PR was laid out in description - now it's just an addition of module which is not used in prophy lib or prophyc app.

File: prophy/utils.py (hunk 9 above)
Comment (kamichal) [reply prophy/utils.py:14]: Good point. `yaml` is much more readable than `json`. I think the reason to implement such exporter is that `JSON` can be used in Java Script out of the box - while `YML` needs some external libraries. I think exposing methods for both would be nice.

File: prophy/utils.py (hunk 6 above)
Comment (mareklabonarski) [reply prophy/utils.py]: here I have a question. 
Why do we have six module in prophy, and cannot use six package?
I have fixed it and defined string_types in prophy.six...

File: prophy/utils.py (hunk 9 above)
Comment (mareklabonarski) [reply prophy/utils.py:14]: Yaml is nice, and good to have here. But since it seems so easy to change json-> yaml (yaml.dump(json_struct)), do we need anything in our library? That can be done externally on the flow...

File: prophy/utils.py (hunk 6 above)
Comment (mareklabonarski) [reply prophy/utils.py]: Seems we have a problem. I cannot define string_types as defined in six package, because tox[check] gives me error of undefined 'basestring'. Looks like we need to use package...

Comment (kamichal) [reply prophy/utils.py]: so do it exactly as in `prophyc/six.py`

File: prophy/utils.py (hunk 9 above)
Comment (kamichal) [reply prophy/utils.py:14]: No it's not needed. Even @aurzenligl did it in this way in his example (`yaml.dump(..)`). However there is one reason to make custom `yaml` frontend: hidden fields. Comments in yaml could help avoiding a confusion. E.g. a unpopulated dynamic array is not rendered in such a json, inactive optional field as well. A comment mentioning that fact could be very desirable.

File: prophy/prophy_six.py [hunk 13]
```diff
@@ -6,7 +6,6 @@
     long = long
     from itertools import ifilter
     xrange = xrange
-    string_types = (basestring, )
```
Comment (kamichal) [inline prophy/prophy_six.py]: Instead of replacing just 10 characters here, you decided to introduce external dependency, rename a shared and change it's usage in 9 files.

File: prophy/utils.py (hunk 11 above)
Comment (mareklabonarski) [reply prophy/utils.py]: changed to exception reraised..

File: prophy/utils.py (hunk 6 above)
Comment (aurzenligl) [reply prophy/utils.py]: > Why do we have six module in prophy, and cannot use six package?

Afair due to fact that six module defines many more compatibility settings than prophy may ever need. There is `future` python project as well which eases the pain of writing in python6:
https://python-future.org/

File: prophy/utils.py (hunk 9 above)
Comment (aurzenligl) [reply prophy/utils.py:14]: If function aims at preparing data structures for further serialization in json/yaml, name: `sanitize_*` could be used. It doesn't mention serialization itself nor serialization format, but it suggests conforming to some stricter standard. Jsonize suggests making and returning json.

**Comment by aurzenligl** [comment]
@mareklabonarski, where are we with this PR, would you like to have it in prophy, or should we close it?

**Comment by mareklabonarski** [comment]
Please let me keep it open and come back to it once all higher priority things are done. 
This MR is unrelated to the recent ones and part of different task, which waits for its time to be finished..
